[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Matthew1172/EquationOfASlime/blob/master/Assignment3.ipynb)

In [146]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn import linear_model

In [147]:
rows = 15
url = "https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv"
df = pd.read_csv(url)

In [148]:
df.head(rows)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [149]:
df.describe()

Temperature °C     Mols KCL     Size nm^3
count     1000.000000  1000.000000  1.000000e+03
mean       500.500000   471.530000  5.086111e+05
std        288.819436   288.482872  4.474838e+05
min          1.000000     1.000000  1.611429e+01
25%        250.750000   226.750000  1.298267e+05
50%        500.500000   459.500000  3.827182e+05
75%        750.250000   710.250000  7.603211e+05
max       1000.000000  1000.000000  1.972127e+06

In [150]:
X, y  = df.iloc[: , :2], df.iloc[: , 2:3]
X

Temperature °C  Mols KCL
0               469       647
1               403       694
2               302       975
3               779       916
4               901        18
..              ...       ...
995             894       847
996             327       982
997             791       213
998             769       553
999             919       452

[1000 rows x 2 columns]

In [151]:
y

Size nm^3
0    6.244743e+05
1    5.779610e+05
2    6.196847e+05
3    1.460449e+06
4    4.325726e+04
..            ...
995  1.545661e+06
996  6.737041e+05
997  3.477543e+05
998  8.684794e+05
999  8.476413e+05

[1000 rows x 1 columns]

In [152]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, train_size=0.9)

In [153]:
lr = LinearRegression().fit(X_train, y_train)
print(lr.predict(np.array([[1, 2]])))
print("Score:", lr.score(X_test, y_test))
print("Coefficients:", lr.coef_)
print("Intercept:", lr.intercept_)

[[-414430.2193772]]
Score: 0.8805773557282204
Coefficients: [[ 868.4983004  1040.04785038]]
Intercept: [-417378.81337835]


C:\Users\pecko\anaconda3\envs\EquationOfASlime\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


$S = -410736 +  869.35852(a) + 1036.00412(b) $
<p>We can see above that by using linear regression, the model achieves an 83.5% accuracy. This means that the values obtained from the line of best fit are about 16% different from the actual difference when comparing how much temperature °C and Mols KCL creates a change in Size nm^3.</p>

In [154]:
cross_val_score(lr, X_test, y_test, cv=6)

array([0.92845546, 0.90823823, 0.88029052, 0.7808221 , 0.89969372,
       0.82026888])

The 6 cross validation score shows the accuracy using linear regression is in the range of 71%-88%.

In [155]:
poly = PolynomialFeatures(degree=2)
X_train = poly.fit_transform(X_train)
X_test = poly.fit_transform(X_test)
reg = linear_model.BayesianRidge()
reg.fit(X_train, y_train)
print("Accuracy:", reg.score(X_test, y_test))
print("Intercept:", reg.intercept_)
print("Coefficients:", reg.coef_)
print("Predict: ", reg.predict(np.array([[1, 2, 3, 5, 6, 7]])))

Accuracy: 1.0
Intercept: 2.0504114218056202e-05
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.58603912e-07 -1.84299243e-11
  2.00000000e+00  2.85714287e-02]
Predict:  [36.20002004]


C:\Users\pecko\anaconda3\envs\EquationOfASlime\lib\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


$y = 1.85481(10^{-5}) + 0(1) + 1.2(10^2)(a) - 1.31005e(10^{-7})(b) - 5.72841(10^{-12})(a^2) + 2(ab) + 2.85714(10^{-2})(b^2)$
<p>The accuracy of the model is 100%. This means that the polynomialfeatures regression is much more accurate than the linear regression. This shows that polynomialfeatures regression achieves higher accuracy than linear regression.</p>

In [156]:
cross_val_score(reg, X_test, y_test, cv=6)

C:\Users\pecko\anaconda3\envs\EquationOfASlime\lib\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\pecko\anaconda3\envs\EquationOfASlime\lib\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\pecko\anaconda3\envs\EquationOfASlime\lib\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\pecko\anaconda3\envs\EquationOfASlime\lib\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y 

array([1., 1., 1., 1., 1., 1.])